[UI-side compositing] css3/masking/clip-path-overflow-hidden-bounds.html fails

kkinnunen-apple opened the pull request:
#### 0ccc458e846190c097901782265d11d9631c234a
<pre>
[UI-side compositing] css3/masking/clip-path-overflow-hidden-bounds.html fails
<a href="https://bugs.webkit.org/show_bug.cgi?id=253575">https://bugs.webkit.org/show_bug.cgi?id=253575</a>
rdar://106115074

Reviewed by Said Abou-Hallawa.

Display list recorder would batch state changes and only emit a state
change item when a draw would need.  This algorithm would hoist state
changes made outside of save/restore blocks to be applied inside:
  ASSERT_TRUE(c.shouldAntialias());
  c.setShouldAntialias(false);
  c.save();
  c.fillRect({..});
  c.restore();

Would become:
  c.save();
  c.setShouldAntialias(false);
  c.fillRect({..});
  c.restore();

For tests like clip-path-overflow-hidden-bounds.html this would fail
on GPUP DOM rendering with layout tests.
The tests would not fail interactively.
This is because the browsers run with device scale &gt; 1, and layout test
runner tests with device scale == 1.
The most noticeable errors due to state change problems were the antialias
toggles. Clipping, for example, toggles these when the device scale == 1
but not otherwise.

* Source/WebCore/platform/graphics/displaylists/DisplayList.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListDrawingContext.cpp:
* Source/WebCore/platform/graphics/displaylists/DisplayListDrawingContext.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::commitRecording):
(WebCore::DisplayList::Recorder::save):
(WebCore::DisplayList::Recorder::restore):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/DisplayListRecorderTests.cpp: Added.
(TestWebKitAPI::createReferenceTarget):
(TestWebKitAPI::createDisplayListTarget):
(TestWebKitAPI::forBoth):
(TestWebKitAPI::checkEqualState):
(TestWebKitAPI::TYPED_TEST_P):

Canonical link: <a href="https://commits.webkit.org/261451@main">https://commits.webkit.org/261451@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a66b7b6a1d114e0f8ddec06bd23b45b59cd321c6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111663 "29 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20799 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/275 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3493 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120400 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115724 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22154 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11879 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3236 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117429 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99600 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104615 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98406 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/179 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45387 "") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13278 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/176 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/87815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13768 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9622 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19213 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52174 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7971 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15758 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
```diff
@@ -61,6 +61,12 @@ Recorder::~Recorder()
     ASSERT(m_stateStack.size() == 1); // If this fires, it indicates mismatched save/restore.
 }
 
+void Recorder::commitRecording()
```
This name is little bit confusing. For me it will mean commit all the recorded display items to the destination context. But actually all it does it ensures the `GraphicsContextState` of destination context is up-to-date before recording any drawing command (and now before recording the `save()` command as well).

kkinnunen-apple replied:
> For me it will mean commit all the recorded display items to the destination context

Almost! There's no destination context in a display list. It commits the items to the display list ! :) 

So sounds like your thinking was almost right, which says the naming is correct.

Ergo it is not the naming that is causing the confusion, you just thought the thing has a destination context when it doesn't.

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
```diff
@@ -265,13 +271,15 @@ void Recorder::drawPattern(ImageBuffer& imageBuffer, const FloatRect& destRect,
 
 void Recorder::restore()
 {
+    appendStateChangeItemIfNecessary();
```
I do not think this is needed. Once `restore() `is recorded, the current `GraphicsContextState` will be popped anyway from the stack of the destination context.

kkinnunen-apple replied:
Nope, you're pushing for introducing the exact same bug. :)
It's also tested in the tests:
```
c.save();
c.setShouldAntialias(false);
c.restore();
c.fillRect({..});
```

Here, if you don't enforce that the state is committed before `restore()`, this will happen:
```
c.save();
c.restore();
c.setShouldAntialias(false);
c.fillRect({..});
```
(The fillRect is the first one that would then commit the state)

shallawa commented on Tools/TestWebKitAPI/Tests/WebCore/DisplayListRecorderTests.cpp:
```diff
@@ -0,0 +1,229 @@
+/*
```
Can't we just add a layout test similar to `displaylists/layer-dispay-list.html` which has the sequence you mentioned in the change log and verify the recorded display list will have `SetState` before `Save`?

kkinnunen-apple replied:
Well, those are good, except
- cannot target specifically all the c++ entry points that are possible
- cannot assert specifically any of the c++ entry points of the graphics context state

E.g. the idea of the test is that whatever the display list does, the corresponding non-display-list state must be the same as the state generated by the display list to the new graphics context.

Like, you cannot test this set of commands:
```
c.setShouldAntialias(false);
```

kkinnunen-apple committed:
[UI-side compositing] css3/masking/clip-path-overflow-hidden-bounds.html fails
https://bugs.webkit.org/show_bug.cgi?id=253575
rdar://106115074

Reviewed by Said Abou-Hallawa.

Display list recorder would batch state changes and only emit a state
change item when a draw would need.  This algorithm would hoist state
changes made outside of save/restore blocks to be applied inside:
  ASSERT_TRUE(c.shouldAntialias());
  c.setShouldAntialias(false);
  c.save();
  c.fillRect({..});
  c.restore();

Would become:
  c.save();
  c.setShouldAntialias(false);
  c.fillRect({..});
  c.restore();

For tests like clip-path-overflow-hidden-bounds.html this would fail
on GPUP DOM rendering with layout tests.
The tests would not fail interactively.
This is because the browsers run with device scale > 1, and layout test
runner tests with device scale == 1.
The most noticeable errors due to state change problems were the antialias
toggles. Clipping, for example, toggles these when the device scale == 1
but not otherwise.

* Source/WebCore/platform/graphics/displaylists/DisplayList.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListDrawingContext.cpp:
* Source/WebCore/platform/graphics/displaylists/DisplayListDrawingContext.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::commitRecording):
(WebCore::DisplayList::Recorder::save):
(WebCore::DisplayList::Recorder::restore):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/DisplayListRecorderTests.cpp: Added.
(TestWebKitAPI::createReferenceTarget):
(TestWebKitAPI::createDisplayListTarget):
(TestWebKitAPI::forBoth):
(TestWebKitAPI::checkEqualState):
(TestWebKitAPI::TYPED_TEST_P):

Canonical link: https://commits.webkit.org/261451@main
